### PR TITLE
[utils] Don't trust Content-length when Content-encoding is present

### LINKED ISSUE #3772: [bilibili.com] unable to --write-subtitles
### Checklist

- [X] I'm asking a question and **not** reporting a bug/feature request
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar questions including closed ones

### Question

I've encountered unexpected behaviour I fail to explain. Could you look into this?

Video: https://www.bilibili.com/video/BV16v41137oJ/?uid=425631367634313133376F4A 

With `--ignore-config` the download works, but with my config file the download fails (gets stuck in an infinite retry loop).

I'm asking because both
- my config file doesn't give me any errors on other sites and
- the error encountered is somewhat strange and seems unrelated to any configuration issues, namely:
`Downloaded 311 bytes, expected 220 bytes.`

I can't fathom what configuration choices could lead to such an error, so I would be thankful if you could take a look. Thanks.

### Verbose log

```shell
yt-dlp --ignore-config -v https://www.bilibili.com/video/BV16v41137oJ/?uid=425631367634313133376F4A
[debug] Command-line config: ['--ignore-config', '-v', 'https://www.bilibili.com/video/BV16v41137oJ/?uid=425631367634313133376F4A']
[debug] Encodings: locale cp1252, fs utf-8, out utf-8, err utf-8, pref cp1252
[debug] yt-dlp version 2022.04.08 [7884ade] (win_exe)
[debug] Python version 3.8.10 (CPython 64bit) - Windows-10-10.0.19043-SP0
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg N-105208-gb24f0c82b3-20220108 (setts), ffprobe N-105208-gb24f0c82b3-20220108
[debug] Optional libraries: brotli, certifi, Cryptodome, mutagen, sqlite, websockets
[debug] Proxy map: {}
[debug] [BiliBili] Extracting URL: https://www.bilibili.com/video/BV16v41137oJ/?uid=425631367634313133376F4A
[BiliBili] 16v41137oJ: Grabbing original ID via API
[BiliBili] 250653958: Downloading webpage
[BiliBili] 250653958: Extracting videos in anthology
[BiliBili] Downloading anthology 250653958 - add --no-playlist to just download video
[download] Downloading playlist: 【纪录片】跳跃的原因 The Reason I Jump（2020）
[BiliBili] playlist 【纪录片】跳跃的原因 The Reason I Jump（2020）: Downloading 1 videos
[download] Downloading video 1 of 1
[debug] [BiliBili] Extracting URL: https://www.bilibili.com/video/BV16v41137oJ?p=1
[BiliBili] 16v41137oJ: Grabbing original ID via API
[BiliBili] 250653958: Downloading webpage
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[BiliBili] 250653958: Extracting videos in anthology
[BiliBili] 250653958: Downloading tags
[debug] Default format spec: bestvideo*+bestaudio/best
[info] 250653958_part1: Downloading 1 format(s): 5+2
[debug] Invoking downloader on "https://upos-hz-mirrorakam.akamaized.net/upgcxcode/74/82/415438274/415438274-1-30064.m4s?e=ig8euxZM2rNcNbdlhoNvNC8BqJIzNbfqXBvEqxTEto8BTrNvN0GvT90W5JZMkX_YN0MvXg8gNEV4NC8xNEV4N03eN0B5tZlqNxTEto8BTrNvNeZVuJ10Kj_g2UB02J0mN0B5tZlqNCNEto8BTrNvNC7MTX502C8f2jmMQJ6mqF2fka1mqx6gqj0eN0B599M=&uipk=5&nbs=1&deadline=1652833956&gen=playurlv2&os=akam&oi=1534581887&trid=900ed95b25f148058a7ab8740963740fu&platform=pc&upsig=ce3778498e2e8766bd0eae72fdff6de0&uparams=e,uipk,nbs,deadline,gen,os,oi,trid,platform&hdnts=exp=1652833956~hmac=f8fe67860023ef3bba6a0a8c27493147d16d12229f9cca43ee0a201c516dd36d&mid=0&bvc=vod&nettype=0&orderid=0,1&agrr=0&bw=190000&logo=80000000"
[debug] File locking is not supported on this platform. Proceeding without locking
[download] Destination: 【纪录片】跳跃的原因 The Reason I Jump（2020） [250653958_part1].f5.mp4
[download]   0.2% of 844.57MiB at 112.92KiB/s ETA 02:07:21
ERROR: Interrupted by user

yt-dlp https://www.bilibili.com/video/BV16v41137oJ/?uid=425631367634313133376F4A
[debug] Command-line config: ['https://www.bilibili.com/video/BV16v41137oJ/?uid=425631367634313133376F4A']
[debug] Portable config "C:\hp\-\setups\+downloaders\yt-dlp\appdir\yt-dlp.conf": ['--verbose', '--format-sort', 'hasvid,ie_pref,quality,res,hdr,codec,size,br,asr,ext,fps,hasaud,lang,source,proto,id', '-f', 'bestvideo*+bestaudio/best', '--paths', 'c:\\Users\\User\\Downloads\\yt', '-o', '.\\%(playlist|)s\\%(playlist_index|)02d%(playlist_index& = |)s%(release_date>%Y-%m-%d,release_timestamp>%Y-%m-%d,upload_date>%Y-%m-%d,timestamp>%Y-%m-%d|#nodate#)s %(alt_title,title)s\\%(title)s [%(extractor)s.%(extractor_key)s.%(id)s] [%(format_id)s].%(ext)s', '-o', 'thumbnail:.\\%(playlist|)s\\%(playlist_index|)02d%(playlist_index& = |)s%(release_date>%Y-%m-%d,release_timestamp>%Y-%m-%d,upload_date>%Y-%m-%d,timestamp>%Y-%m-%d|#nodate#)s %(alt_title,title)s\\thumbs\\thumb.%(ext)s', '-o', 'subtitle:.\\%(playlist|)s\\%(playlist_index|)02d%(playlist_index& = |)s%(release_date>%Y-%m-%d,release_timestamp>%Y-%m-%d,upload_date>%Y-%m-%d,timestamp>%Y-%m-%d|#nodate#)s %(alt_title,title)s\\subs\\%(alt_title,title)s [%(extractor)s.%(extractor_key)s.%(id)s] [%(format_id)s].%(ext)s', '-o', 'infojson:.\\%(playlist|)s\\%(playlist_index|)02d%(playlist_index& = |)s%(release_date>%Y-%m-%d,release_timestamp>%Y-%m-%d,upload_date>%Y-%m-%d,timestamp>%Y-%m-%d|#nodate#)s %(alt_title,title)s\\!nfo\\%(alt_title,title)s [%(extractor)s.%(extractor_key)s.%(id)s] [%(format_id)s].%(ext)s', '-o', 'link:.\\%(playlist|)s\\%(playlist_index|)02d%(playlist_index& = |)s%(release_date>%Y-%m-%d,release_timestamp>%Y-%m-%d,upload_date>%Y-%m-%d,timestamp>%Y-%m-%d|#nodate#)s %(alt_title,title)s\\!nfo\\%(title)s [%(extractor)s.%(extractor_key)s.%(id)s] [%(format_id)s].%(ext)s', '-o', 'description:.\\%(playlist|)s\\%(playlist_index|)02d%(playlist_index& = |)s%(release_date>%Y-%m-%d,release_timestamp>%Y-%m-%d,upload_date>%Y-%m-%d,timestamp>%Y-%m-%d|#nodate#)s %(alt_title,title)s\\!nfo\\%(title)s [%(extractor)s.%(extractor_key)s.%(id)s] [%(format_id)s].%(ext)s', '-o', 'annotation:.\\%(playlist|)s\\%(playlist_index|)02d%(playlist_index& = |)s%(release_date>%Y-%m-%d,release_timestamp>%Y-%m-%d,upload_date>%Y-%m-%d,timestamp>%Y-%m-%d|#nodate#)s %(alt_title,title)s\\!nfo\\%(title)s [%(extractor)s.%(extractor_key)s.%(id)s] [%(format_id)s].%(ext)s', '-o', 'pl_infojson:.\\%(playlist|)s\\playlist.%(ext)s', '-o', 'pl_thumbnail:.\\%(playlist|)s\\folder.%(ext)s', '-o', 'pl_description:.\\%(playlist|)s\\playlist description.%(ext)s', '--output-na-placeholder', '~~~', '--write-info-json', '--write-description', '--write-annotations', '--no-overwrites', '--no-post-overwrites', '--merge-output-format', 'mkv', '--remux-video', 'mkv', '--write-subs', '--write-auto-subs', '--all-subs', '--sub-langs', 'all', '--sub-format', 'all/ttml/vtt/best', '--write-thumbnail', '--write-all-thumbnails', '--add-metadata', '--embed-chapters', '--add-chapters', '--write-link', '--write-url-link', '--write-desktop-link', '--no-check-certificate', '--retries', 'infinite', '--geo-bypass', '--no-clean-infojson', '--write-comments', '--audio-multistreams', '--no-playlist']
[debug] Encodings: locale cp1252, fs utf-8, out utf-8, err utf-8, pref cp1252
[debug] yt-dlp version 2022.04.08 [7884ade] (win_exe)
[debug] Python version 3.8.10 (CPython 64bit) - Windows-10-10.0.19043-SP0
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg N-105208-gb24f0c82b3-20220108 (setts), ffprobe N-105208-gb24f0c82b3-20220108
[debug] Optional libraries: brotli, certifi, Cryptodome, mutagen, sqlite, websockets
[debug] Proxy map: {}
[debug] [BiliBili] Extracting URL: https://www.bilibili.com/video/BV16v41137oJ/?uid=425631367634313133376F4A
[BiliBili] 16v41137oJ: Grabbing original ID via API
[BiliBili] 250653958: Downloading webpage
[BiliBili] Downloading just video 250653958 because of --no-playlist
[debug] Sort order given by user: hasvid, ie_pref, quality, res, hdr, codec, size, br, asr, ext, fps, hasaud, lang, source, proto, id
[debug] Formats sorted by: hasvid, ie_pref, quality, res, hdr, vcodec, acodec, filesize, fs_approx, tbr, vbr, abr, asr, vext, aext, fps, hasaud, lang, source, proto, id
[BiliBili] 250653958: Downloading tags
[debug] Downloading subtitles: danmaku
[BiliBili] 250653958: Extracting comments from page 1
[BiliBili] Extracted 0 comments
[info] 250653958_part1: Downloading 1 format(s): 5+2
[info] Video description is already present
[info] Writing video subtitles to: c:\Users\User\Downloads\yt\2021-09-27 【纪录片】跳跃的原因 The Reason I Jump（2020）\【纪录片】跳跃的原因 The Reason I Jump（2020） [BiliBili.BiliBili.250653958_part1] [5+2].danmaku.xml
[debug] Invoking downloader on "https://comment.bilibili.com/415438274.xml"
[download] Resuming download at byte 311
[download] Unable to resume
[debug] File locking is not supported on this platform. Proceeding without locking
[download] Destination: c:\Users\User\Downloads\yt\2021-09-27 【纪录片】跳跃的原因 The Reason I Jump（2020）\【纪录片】跳跃的原因 The Reason I Jump（2020） [BiliBili.BiliBili.250653958_part1] [5+2].danmaku.xml
[download] 141.4% of 220.00B at Unknown speed ETA Unknown[download] Got server HTTP error: Downloaded 311 bytes, expected 220 bytes. Retrying (attempt 1 of inf) ...
[download] Unable to resume
[download] Destination: c:\Users\User\Downloads\yt\2021-09-27 【纪录片】跳跃的原因 The Reason I Jump（2020）\【纪录片】跳跃的原因 The Reason I Jump（2020） [BiliBili.BiliBili.250653958_part1] [5+2].danmaku.xml
[download] 141.4% of 220.00B at Unknown speed ETA Unknown[download] Got server HTTP error: Downloaded 311 bytes, expected 220 bytes. Retrying (attempt 2 of inf) ...
[download] Unable to resume
[download] Destination: c:\Users\User\Downloads\yt\2021-09-27 【纪录片】跳跃的原因 The Reason I Jump（2020）\【纪录片】跳跃的原因 The Reason I Jump（2020） [BiliBili.BiliBili.250653958_part1] [5+2].danmaku.xml
[download] 141.4% of 220.00B at 19.81KiB/s ETA 00:00[download] Got server HTTP error: Downloaded 311 bytes, expected 220 bytes. Retrying (attempt 3 of inf) ...
[download] Unable to resume
[download] Destination: c:\Users\User\Downloads\yt\2021-09-27 【纪录片】跳跃的原因 The Reason I Jump（2020）\【纪录片】跳跃的原因 The Reason I Jump（2020） [BiliBili.BiliBili.250653958_part1] [5+2].danmaku.xml
[download] 141.4% of 220.00B at Unknown speed ETA Unknown[download] Got server HTTP error: Downloaded 311 bytes, expected 220 bytes. Retrying (attempt 4 of inf) ...
[download] Unable to resume
[download] Destination: c:\Users\User\Downloads\yt\2021-09-27 【纪录片】跳跃的原因 The Reason I Jump（2020）\【纪录片】跳跃的原因 The Reason I Jump（2020） [BiliBili.BiliBili.250653958_part1] [5+2].danmaku.xml
[download] 141.4% of 220.00B at 19.82KiB/s ETA 00:00[download] Got server HTTP error: Downloaded 311 bytes, expected 220 bytes. Retrying (attempt 5 of inf) ...
[download] Unable to resume
[download] Destination: c:\Users\User\Downloads\yt\2021-09-27 【纪录片】跳跃的原因 The Reason I Jump（2020）\【纪录片】跳跃的原因 The Reason I Jump（2020） [BiliBili.BiliBili.250653958_part1] [5+2].danmaku.xml
[download] 141.4% of 220.00B at Unknown speed ETA Unknown[download] Got server HTTP error: Downloaded 311 bytes, expected 220 bytes. Retrying (attempt 6 of inf) ...
[download] Unable to resume
[download] Destination: c:\Users\User\Downloads\yt\2021-09-27 【纪录片】跳跃的原因 The Reason I Jump（2020）\【纪录片】跳跃的原因 The Reason I Jump（2020） [BiliBili.BiliBili.250653958_part1] [5+2].danmaku.xml
[download] 141.4% of 220.00B at 19.43KiB/s ETA 00:00[download] Got server HTTP error: Downloaded 311 bytes, expected 220 bytes. Retrying (attempt 7 of inf) ...
[download] Unable to resume
[download] Destination: c:\Users\User\Downloads\yt\2021-09-27 【纪录片】跳跃的原因 The Reason I Jump（2020）\【纪录片】跳跃的原因 The Reason I Jump（2020） [BiliBili.BiliBili.250653958_part1] [5+2].danmaku.xml
[download] 141.4% of 220.00B at Unknown speed ETA Unknown[download] Got server HTTP error: Downloaded 311 bytes, expected 220 bytes. Retrying (attempt 8 of inf) ...
[download] Unable to resume
[download] Destination: c:\Users\User\Downloads\yt\2021-09-27 【纪录片】跳跃的原因 The Reason I Jump（2020）\【纪录片】跳跃的原因 The Reason I Jump（2020） [BiliBili.BiliBili.250653958_part1] [5+2].danmaku.xml
[download] 141.4% of 220.00B at 19.48KiB/s ETA 00:00[download] Got server HTTP error: Downloaded 311 bytes, expected 220 bytes. Retrying (attempt 9 of inf) ...
```

### PR DESCRIPTION
**Pull request by felixonmars**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Our YoutubeDLHandler is doing auto-decompression, so that we cannot compare Content-length to the size of extracted data. Although HttpFD has specified Youtubedl-no-compression, the web server may still ignore this and return deflated data (as the situation in #3772).

This change attempts to retain the Content-encoding header for later comparison in HttpFD so that we could discard data_len when the web server returns compressed data. I didn't find the reason why this header was previously removed but didn't yet find a counter example either.

Fixes #3772 in my experiments.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [x] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by pukkandan** [comment]
cc @coletdjnz

**Comment by coletdjnz** [comment]
note for self: check against urllib3/requests too

File: yt_dlp/downloader/http.py [hunk 1]
```diff
@@ -213,6 +213,11 @@ def close_stream():
         def download():
             data_len = ctx.data.info().get('Content-length', None)
 
+            if ctx.data.info().get('Content-encoding', None):
```
Comment (dirkf) [inline yt_dlp/downloader/http.py:216]: `None` is default?
```suggestion
            if ctx.data.info().get('Content-encoding'):
```

**Comment by dirkf** [comment]
>... I didn't find the reason why this header was previously removed but didn't yet find a counter example either.

When I worked on this code I inferred that it was trying to handle nested encodings, or perhaps more realistically to avoid decoding an already decoded contents. The old yt-dl code was like this:
```
if resp.headers.get('Content-encoding', '') == method1:
    #

File: yt_dlp/downloader/http.py (hunk 1 above)
Comment (felixonmars) [reply yt_dlp/downloader/http.py:216]: Yes, I'm just trying to keep the same coding style as data_len is also adding this `, None`. I'll try to remove both then.

**Comment by felixonmars** [comment]
> When I worked on this code I inferred that it was trying to handle nested encodings, or perhaps more realistically to avoid decoding an already decoded contents.

I see. An alternative approach would be moving the info into a custom header like "Youtubedl-decompressed" and check that in HttpFD instead.

**Comment by coletdjnz** [comment]
urllib3/requests appear to retain the content-encoding header after decompression too, so this should be fine in that regard (needed to check due to #2861 and #3668).

> When I worked on this code I inferred that it was trying to handle nested encodings, or perhaps more realistically to avoid decoding an already decoded contents.

Shouldn't be an issue as the data should only be tried to be decoded once (i.e. content-encoding read once for whole decoding process, including multi decoding). I don't see us needing anyway of checking if the content is decoded or not either (and in the case it is in an unsupported encoding, we can catch that elsewhere), so including the header even when decoded should be fine.

**Review by coletdjnz — APPROVED** [review]
Networking changes look fine to me.

https://github.com/yt-dlp/yt-dlp/issues/3772 seems to work for me too now assuming you don't use `--test` (@pukkandan not sure if that is a concern)

File: yt_dlp/downloader/http.py [hunk 2]
```diff
@@ -211,7 +211,12 @@ def close_stream():
                 ctx.stream = None
 
         def download():
-            data_len = ctx.data.info().get('Content-length', None)
+            data_len = ctx.data.info().get('Content-length')
+
+            if ctx.data.info().get('Content-encoding'):
+                # Content-encoding is present, Content-length is not reliable anymore as we are
+                # doing auto decompression.
```
Comment (coletdjnz) [inline yt_dlp/downloader/http.py:218]: Add a reference back to this pr or the issue

**Review by coletdjnz — CHANGES_REQUESTED** [review]

**Comment by pukkandan** [comment]
> assuming you don't use `--test` (@pukkandan not sure if that is a concern)

Let's not worry about that for now

Make sure to port the change to the networking PR

**Pull request merged**